### ISSUE TITLE
Known issues in `add-geo-metadata-to-table.R`

### ISSUE THREAD

**Issue by bswhite** [mc2-center/csbc-pson-dcc#43]
Plenty of issues with the new script I just checked in to download GEO metadata and add to a table, but it gets the job done (enough) for now:

- [ ] 1. It's slow because I've used serial for loops.
- [ ] 2. It confuses accessing GEO annotations with standardizing vocabularies. e.g., it translates some taxon to our ontologies (but not all). Probably this shouldn't be in this script.
- [ ] 3. It doesn't properly handle multi-values/JSON in taxon. 
- [ ] 4. It doesn't properly handle quotes within columns. 

Because of 3 and 4, you will need to manually edit output before uploading to Synapse. e.g.,
`\"\"` -> `""`
`\\"\"` -> `""`

**Comment by bswhite** [comment]
Also, I bet this could be extended to download GSE and SRA links. Though, unlike the other annotations, these might involve some interpretation / translation.

**Comment by jaeddy** [comment]
Excellent! Thanks, @bswhite! I'll take a closer look at the new code tomorrow. 

Here's what I've managed to capture (for the first 9 BioProject IDs of 199 for our current set of PMIDs — i.e., 6 PubMed IDs -> 9 BioProject IDs -> 268 samples):
https://www.synapse.org/#!Synapse:syn22015979/tables/

I need to do some additional debugging to get through the full list of matched BioProjects... then I'll try to suss out whether or not we actually need to query any DBs other than SRA to get what we need.

**Comment by bswhite** [comment]
This looks great, but it looks like it's still missing overall design.

The first row in your table is from
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103242
and should have overall design
Transcriptome profiles of tamoxifen sensitive MCF-7 infected with pLenti-CMV-hygro-GFP or pLento-HA-EZH2 were sequenced in at least duplicates using Illumina HiSeq 3000.

(study_abstract is called "summary" on the GEO website)

Unless there is another option you can add to whatever library/approach you are using, I think we still have a need to use multiple tools to get all of the fields of interest (at either the dataset or individual file level). Not a barrier, but annoying.

I had been getting similar sample-level information using the get-geo-annotation.R script. It certainly doesn't have the richness in the links to other databases (which I've been using other scripts for) or the dataset-level information. I stashed the output from that script here:
https://www.synapse.org/#!Synapse:syn21788461

As a point of comparison to the above, the corresponding annotations for GSE103242 are:
https://www.synapse.org/#!Synapse:syn21788627

By comparing the two, I see that the sample_attribute field above is a concatenation of characteristics_ch1 and source_name_ch1 from the get-geo script.

**Comment by jaeddy** [comment]
I thought the lack of overall design was so odd, but I found it! It's actually embedded in the `study_abstract`. 

Yeah, the `sample_attribute` is returned as a list of `$TAG` and `$VALUE` elements, so I concatenated so rows would line up. We could potentially parse them out into separate columns at some point, but based on my own experience submitting to GEO, I'd be shocked if field names were standardized.

I'll add my code to the repo and maybe ask Andrew to see if he can get it to run for all of the matched BioProjects. We can decide later what (if anything) to do with the table.
